### PR TITLE
Implement Swap logic for concentrated liquidity

### PR DESCRIPTION
**Pull request by mattverse**
This PR adds the swapping logic(mutating the pool state) and adds tests to check whether the pool was correctly mutated

### CODE REVIEW

File: x/concentrated-liquidity/pool.go [hunk 1]
```diff
@@ -85,7 +95,12 @@ type SwapState struct {
 }
 
 // this only works on a single directional trade, will implement bi directional trade in next milestone
-func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, swapFee sdk.Dec, minPrice, maxPrice sdk.Dec, poolId uint64) (newTokenIn, tokenOut sdk.Coin, err error) {
+func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	swapFee sdk.Dec,
+	minPrice, maxPrice sdk.Dec,
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:102]: we can get rid of max and min price now right?

File: x/concentrated-liquidity/pool.go [hunk 2]
```diff
@@ -68,8 +69,17 @@ func (p Pool) GetTotalShares() sdk.Int {
...
 
-func (p Pool) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coins, tokenOutDenom string, swapFee sdk.Dec) (tokenOut sdk.Coin, err error) {
-	return sdk.Coin{}, nil
+func (k Keeper) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDenom string, swapFee sdk.Dec, minPrice, maxPrice sdk.Dec, poolId uint64) (tokenOut sdk.Coin, err error) {
+	tokenInCoin, newLiquidity, newCurrentTick, newCurrentSqrtPrice, err := k.CalcOutAmtGivenIn(ctx, tokenIn, tokenOutDenom, swapFee, minPrice, maxPrice, poolId)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	fmt.Println("=====")
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:78]:
```suggestion
```

File: x/concentrated-liquidity/pool.go [hunk 3]
```diff
@@ -141,41 +156,52 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDen
 		amountCalculated:         sdk.ZeroDec(),
 		sqrtPrice:                curSqrtPrice,
 		tick:                     priceToTick(curSqrtPrice.Power(2)),
+		liquidity:                p.Liquidity,
 	}
 
+	fmt.Println("====cur sqrt price")
+	fmt.Println(curSqrtPrice.String())
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:163]:
```suggestion
```

File: x/concentrated-liquidity/pool.go [hunk 4]
```diff
@@ -141,41 +156,52 @@ func (k Keeper) CalcOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDen
...
+		fmt.Println("===sqrt price")
+		fmt.Println(sqrtPrice.String())
+		fmt.Println("===next sqrt price")
+		fmt.Println(nextSqrtPrice.String())
 		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.Sub(amountIn)
 		swapState.amountCalculated = swapState.amountCalculated.Add(amountOut)
 
+		// if we have moved to the next tick,
 		if nextSqrtPrice.Equal(sqrtPrice) {
+			fmt.Println("this is inside m=== == = = ")
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:193]:
```suggestion
```

File: x/concentrated-liquidity/pool.go [hunk 5]
```diff
@@ -254,15 +286,19 @@ func (k Keeper) CalcInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coin, tokenInDen
...
 	// TODO: This should be GT 0 but some instances have very small remainder
 	// need to look into fixing this
+	iteration := 0
 	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 6)) {
+		iteration++
 		nextTick, _ := k.NextInitializedTick(ctx, poolId, swapState.tick.Int64(), zeroForOne)
+		// fmt.Println("---next tick")
+		// fmt.Println(nextTick)
 		// TODO: we can enable this error checking once we fix tick initialization
 		// if !ok {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:296]: we can enable this check right?

File: x/concentrated-liquidity/pool.go [hunk 6]
```diff
@@ -298,5 +333,19 @@ func (k Keeper) CalcInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coin, tokenInDen
 			swapState.tick = priceToTick(sqrtPrice.Power(2))
 		}
 	}
-	return sdk.NewCoin(tokenInDenom, swapState.amountCalculated.RoundInt()), sdk.NewCoin(tokenOut.Denom, tokenOut.Amount), nil
+	fmt.Println("==num of iteration")
+	fmt.Println(iteration)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:337]:
```suggestion
```

File: x/concentrated-liquidity/pool.go [hunk 7]
```diff
@@ -254,15 +286,19 @@ func (k Keeper) CalcInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coin, tokenInDen
 
 	// TODO: This should be GT 0 but some instances have very small remainder
 	// need to look into fixing this
+	iteration := 0
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:289]:
```suggestion
```

**Review by czarcas7ic — APPROVED** [review]

File: x/concentrated-liquidity/pool.go (hunk 5 above)
Comment (mattverse) [reply x/concentrated-liquidity/pool.go:297]: enabled

File: x/concentrated-liquidity/pool.go (hunk 1 above)
Comment (mattverse) [reply x/concentrated-liquidity/pool.go:100]: this is to be done in upcoming PR

**Pull request merged**